ccpp/data/GFS_typedefs.F90

```diff
@@ -5394,29 +5420,64 @@ subroutine control_initialize (Model, nlunit, fn_nml, me, master, &
-    !
-    ! from module ozinterp
-    if (Model%ntoz>0) then
-       if (Model%oz_phys) then
-          levozp   = 80
-          oz_coeff = 4
-       else if (Model%oz_phys_2015) then
-          levozp   = 53
-          oz_coeff = 6
-       else
-          write(*,*) 'Logic error, ntoz>0 but no ozone physics selected'
-          stop
-       end if
+    !--- NRL ozone physics
+    if (Model%ntoz > 0) then
+       ! Get dimensions from data file
+       open(unit=Model%kozpl,file='global_o3prdlos.f77', form='unformatted', convert='big_endian')
+       read (Model%kozpl) Model%oz_coeff, Model%latsozp, Model%levozp, Model%timeoz
+       rewind(Model%kozpl)
+       if (Model%me == Model%master) then
+          write(*,*) 'Reading in o3data from global_o3prdlos.f77 '
+          write(*,*) '      oz_coeff = ', Model%oz_coeff
+          write(*,*) '       latsozp = ', Model%latsozp
+          write(*,*) '        levozp = ', Model%levozp
+          write(*,*) '        timeoz = ', Model%timeoz
+       endif
+       ! Allocate space
+       allocate (Model%oz_lat(Model%latsozp))
+       allocate (Model%oz_pres(Model%levozp))
+       allocate (Model%po3(Model%levozp))
+       allocate (Model%oz_time(Model%timeoz+1))
+       allocate (Model%ozplin(Model%latsozp,Model%levozp,Model%oz_coeff,Model%timeoz))
+       !
+       allocate(oz_lat4(Model%latsozp), oz_pres4(Model%levozp), oz_time4(Model%timeoz+1))
+       read (Model%kozpl) Model%oz_coeff, Model%latsozp, Model%levozp, Model%timeoz, oz_lat4, oz_pres4, oz_time4
+
+       ! Store
+       Model%oz_pres(:) = oz_pres4(:)
+       Model%po3(:)     = log(100.0*Model%oz_pres(:)) ! from mb to ln(Pa)
+       Model%oz_lat(:)  = oz_lat4(:)
+       Model%oz_time(:) = oz_time4(:)
+       deallocate(oz_lat4, oz_pres4, oz_time4)
+       !
+       allocate(tempin(Model%latsozp))
+       do i1=1,Model%timeoz
+          do i2=1,Model%oz_coeff
+             do i3=1,Model%levozp
+                READ(Model%kozpl) tempin
+                Model%ozplin(:,i3,i2,i1) = tempin(:)
+             enddo
+          enddo
+       enddo
+       deallocate(tempin)
+       close(Model%kozpl)
     else
-       if (Model%oz_phys .or. Model%oz_phys_2015) then
-          write(*,*) 'Logic error, ozone physics are selected, but ntoz<=0'
-          stop
+       !--- Diagnostic ozone
+       rewind (Model%kozc)
+       read (Model%kozc,end=101) Model%latsozc, Model%levozc, Model%timeozc, blatc4
+101    if (Model%levozc  < 10 .or. Model%levozc > 100) then
+          rewind (Model%kozc)
+          Model%levozc  = 17
+          Model%latsozc = 18
+          Model%blatc   = -85.0
        else
-          levozp   = 1
-          oz_coeff = 1
-       end if
+          Model%blatc   = blatc4
+       endif
+       Model%latsozp   = 2
+       Model%levozp    = 1
+       Model%timeoz    = 1
+       Model%oz_coeff  = 0
+       Model%dphiozc   = -(Model%blatc+Model%blatc)/(Model%latsozc-1)
     end if
```

Comment:
If we're moving the read_o3data code here, why not also the setindxoz code that also gets executed in the init phase? Since there are very similar 'setindxX` routines in ozinterp, h2ointerp, aerinterp, iccninterp, couldn't a common subroutine/function in here be used for all of them to reduce code replication?

Reply:
Good point. I only moved over the data input from the _init step, not the other pieces.

Reply:
@grantfirl 
A point worth mentioning. 
For a scheme to be stateless we can't allocate data on the HEAP during the physics initialization. To address this we can allocate the data on the host-side.

So in the case where a scheme has it's own external data, to achieve "statelessness", one could either a) allocate space here on the host side, and then read in during the physics initialization phase, or b) allocate and read in the data here. 
I chose the later to avoid having to open the file twice, and to also avoid passing the file identifier information into the physics. However, this raises the question you brought up "Why have any scheme initialization in the physics? Instead do it on the host?"